app/file.go

```diff
@@ -545,28 +545,18 @@ type UploadFileTask struct {
 
 func (t *UploadFileTask) init(a *App) {
 	t.buf = &bytes.Buffer{}
-	t.maxFileSize = *a.Config().FileSettings.MaxFileSize
-	t.limit = *a.Config().FileSettings.MaxFileSize
+	if t.ContentLength > 0 {
+		t.limit = t.ContentLength
+	} else {
+		t.limit = t.maxFileSize
+	}
 
 	t.fileinfo = model.NewInfo(filepath.Base(t.Name))
 	t.fileinfo.Id = model.NewId()
 	t.fileinfo.CreatorId = t.UserId
 	t.fileinfo.CreateAt = t.Timestamp.UnixNano() / int64(time.Millisecond)
 	t.fileinfo.Path = t.pathPrefix() + t.Name
 
-	// Prepare to read ContentLength if it is known, otherwise limit
-	// ourselves to MaxFileSize. Add an extra byte to check and fail if the
-	// client sent too many bytes.
-	if t.ContentLength > 0 {
-		t.limit = t.ContentLength
-		// Over-Grow the buffer to prevent bytes.ReadFrom from doing it
-		// at the very end.
-		t.buf.Grow(int(t.limit + 1 + bytes.MinRead))
-	} else {
-		// If we don't know the upload size, grow the buffer somewhat
-		// anyway to avoid extra reslicing.
-		t.buf.Grow(UploadFileInitialBufferSize)
```

Comment:
Does it make sense to always grow up to `UploadFileInitialBufferSize` to avoid the penalty for small files? Trying to find a middle ground between allocating too much and not allocating at all.

Reply:
Yes, I believe we can keep some form of this although it's hard to say what the optimal value would be since it really depends on the use-case. 

If we take our community as an example, most of the uploaded files (86%) are below 1MB in size.

I guess we could allocate for exactly `content-length` bytes if its value is less than this defined constant (max bytes of pre-allocation we tolerate) and otherwise allocate the max.

Reply:
Sounds good to me :+1:

Reply:
All that preallocation code was there to avoid "heap fragmentation", as customers were reporting multi-Gb VM footprint, and I was able to narrow the cause down to the seemingly fragmented heap. 

Starting with 1Mb appears very reasonable anyway, in practical terms all mattermost app uploads come as mime multipart, so there's no content-length available on individual files.